cpp/ql/test/experimental/query-tests/Security/CWE/CWE-078/WordexpTainted.expected

```diff
@@ -0,0 +1,11 @@
+edges
+| test.cpp:23:20:23:23 | argv | test.cpp:29:13:29:20 | (const char *)... |
+| test.cpp:23:20:23:23 | argv | test.cpp:29:13:29:20 | filePath |
+nodes
+| test.cpp:23:20:23:23 | argv | semmle.label | argv |
+| test.cpp:29:13:29:20 | (const char *)... | semmle.label | (const char *)... |
+| test.cpp:29:13:29:20 | filePath | semmle.label | filePath |
+subpaths
+#select
+| test.cpp:29:13:29:20 | (const char *)... | test.cpp:23:20:23:23 | argv | test.cpp:29:13:29:20 | (const char *)... | Using user-supplied data in a `wordexp` command, without disabling command substitution, can make code vulnerable to command injection. |
```

Comment:
If I read this right, CodeQL identifies two different sinks at the same position and therefore generates two alerts, doesn't it?
I expected only one alert, is there a way to fix this?

Reply:
That's usually due to implicit conversions. In this case, I think it's because of the implicit conversion from `char*` to `const char*`. You can try replacing
```ql
fc.getArgument(0) = sink.asExpr() and
```
with
```ql
fc.getArgument(0).getFullyConverted() = sink.asConvertedExpr() and
```

Alternatively, you can make the `filePath` in your example a `const char*` to avoid the implicit conversion.

Reply:
In general, when looking for sinks, should one always use the  `foo.getFullyConverted() = sink.asConvertedExpr()` pattern to get unique results?
In Java I always just wrote `sink.asExpr()` :D

Reply:
Note that the code scanning / LGTM interfaces will collapse the duplicate results somewhat, so while its nice to fix them they should not affect end users too badly.

Reply:
> In general, when looking for sinks, should one always use the foo.getFullyConverted() = sink.asConvertedExpr() pattern to get unique results?
In Java I always just wrote sink.asExpr() :D

That's because Java represents conversions differently from how it's done in C/C++ (AFAIK, they're more explicit in Java). Since we can't do that for C/C++ since basically any expression would be hidden behind layers of nested implicit and explicit conversions.

So in C/C++, when you have snippet like:
```cpp
void test(void* ptr) {
  int* x = (int*)ptr;
}
```
and you ask for the right-hand side of the initialization, you'll get back `ptr` (and _not_ `(int*)ptr`). You can then ask for the fully converted expression of `ptr` and get back `(int*)ptr`.

As @geoffw0 said, this isn't a super big deal as Code Scanning (and LGTM) will collapse some of these results together.

Reply:
Okay, then I'll leave the query as-is, since this is just in a test and won't be a problem for end users :)